### PR TITLE
Adds the AFK auto cryo removal system (preference setting for players)

### PR DESCRIPTION
**Pull request by farie82**
**What does this PR do:**
Adds the AFK subsystem. It does the following:
1. Gives you a warning after a configurable time saying that you will be cryod after X minutes and despawned after Y minutes.
2. Cryos you after X minutes (configurable). If you're in a secure location (cells etc) you will be despawned instantly
3. Despawns you after Y minutes.

This system is fully voluntarily. It's a preference in the game preference tab that is by default set to no.

The system will ignore players at the following situations:
1. The player doesn't have the preference set.
2. The player has a special role (antag)
3. The player is off station
4. The player is SSD/unconcious/dead
5. The player is in parts of security (everywhere where prisoners can roam around free)
6. The player is restrained
7. The player doesn't have a job assigned to him (special spawned)

Requires a SQL change.

**Changelog:**
:cl:
add: Adds the AFK auto cryo system. By default it won't affect players unless they activate it themselves by setting the preference in their game preferences tab.
/:cl:

### CODE REVIEW

**Comment by Ty-Omaha** [comment]
>Cryos you after X minutes (configurable). If you're in a secure location (cells etc) you will be despawned instantly

This will confuse a lot of security and cause unnecessary "WHERE IS X?" panic, a blacklist for secure areas would be better

This should have an antag blacklist: nukeops, shadowthralls, blob, etc It wouldn't be good for them to get despawned.

As well as a restrained and unconscious blacklist

And finally a station Z-level whitelist, so people can't use it to game for a free TP back to the station, in addition to special event and CC characters being affected

**Comment by farie82** [comment]
> > Cryos you after X minutes (configurable). If you're in a secure location (cells etc) you will be despawned instantly
> 
> This will confuse a lot of security and cause unnecessary "WHERE IS X?" panic, a blacklist for secure areas would be better
This is already a thing in perma brig. However I can make it so that it'll tell security on the comms that one person got cryod.

> This should have an antag blacklist: nukeops, shadowthralls, blob, etc It wouldn't be good for them to get despawned.
It's a voluntary option. Default is off. I doubt a lot of people will use this. And the ones using it will know the consequences of the system. I will add logging for admins so they will get notified when an antag leaves.

> As well as a restrained and unconscious blacklist
Sounds logical. Will add it.

> And finally a station Z-level whitelist, so people can't use it to game for a free TP back to the station, in addition to special event and CC characters being affected
I'll just keep it to Z lvl 1 for the Z levels. Same response to the special event characters as for the antags. It won't affect Aghosting admins (Will be handled by the unconcious check).

**Comment by Ty-Omaha** [comment]
>It's a voluntary option. Default is off. I doubt a lot of people will use this. And the ones using it will know the consequences of the system. I will add logging for admins so they will get notified when an antag leaves.

People aren't going to remember "better turn off my pref!" when they are excited as a nukeop or shadowling. The scenario of a nukeop, wizard, team antag being auto cryo'd should be mechanically prevented. 

There's no point in adding a log _just_ so we can punish someone for forgetting about toggling off a pref when we can mechanically prevent it.

**Comment by farie82** [comment]
> > It's a voluntary option. Default is off. I doubt a lot of people will use this. And the ones using it will know the consequences of the system. I will add logging for admins so they will get notified when an antag leaves.
> 
> People aren't going to remember "better turn off my pref!" when they are excited as a nukeop or shadowling. The scenario of a nukeop, wizard, team antag being auto cryo'd should be mechanically prevented.
> 
> There's no point in adding a log _just_ so we can punish someone for forgetting about toggling off a pref when we can mechanically prevent it.

This situation happens if both the nuke ops and the wizard stand AFK for X amount of minutes on the station (with the requested changes applied). Both also don't have jobs so they are not viable for the system.

However if you really want antags to not be affected by this then I'll add it.

**Comment by farie82** [comment]
The requested features have been added

**Comment by pinatacolada** [comment]
command might be better to be exempted too

**Comment by farie82** [comment]
@pinatacolada wont include that. Its part of the reason why I made the PR

**Comment by pinatacolada** [comment]
it will lead to people not knowing where the cmo went or thinking the captain is dead, because they are supposed to announce they're are going and now they're just not responding. this happened last night on normal cryos

**Comment by pinatacolada** [comment]
yet most of the remaining command failed to notice both of those and kept shouting for the cryod captain for 15 minutes

**Comment by farie82** [comment]
Then by that argument we should also remove cryosleep. Since people can use it without telling others.
Sorry but I wont include that restraint

**Comment by Lunaimon** [comment]
Just make the cryo message really big if it's a command member. Then people won't miss it.

**Comment by Citinited** [comment]
Perhaps a better solution would be to announce the person's role, as well as name, when they get despawned?

e.g. `Biff McGriff (Captain) has entered long-term storage.`

**Comment by farie82** [comment]
Thats something for another PR. This one doesnt change how cryo works

**Comment by MrMagolor** [comment]
Isn't part of the RULES that you have to ahelp and inform command when you cryo as a command member? This seems counter-productive to that.

**Comment by farie82** [comment]
> Isn't part of the RULES that you have to ahelp and inform command when you cryo as a command member? This seems counter-productive to that.

Exactly the reason why I made it.
Some people cannot do that due to medical conditions or otherwise. For example they fall asleep out of nowhere.
Currently those people will get banned or can't play the roles. With this PR I want to give those people the option of playing again but with these restrictions.

The system will send messages to admins when cryoing and such happens. So it's kinda the same

File: code/controllers/subsystem/afk.dm [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+/datum/controller/subsystem/afk/fire()
+	var/list/toRemove = list()
+	for(var/mob/living/carbon/human/H in GLOB.living_mob_list)
+		if(!H.ckey) // Useless non ckey creatures
+			continue
+		
+		var/turf/T 
+		// Only players and players with the AFK watch enabled
+		// No dead, unconcious, restrained, people without jobs, people on other Z levels than the station or antags
+		if(H.client == null || !H.client.prefs.afk_watch || !H.mind || \
```
Comment (Couls) [inline code/controllers/subsystem/afk.dm]: Why use H.client == null instead of !H.client?

Comment (farie82) [reply code/controllers/subsystem/afk.dm]: I think I was coding a lot in Progress that day. Where you explicitly need to say it's null. I'll change it to be more in line with the byond code

**Pull request merged**